sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -213,15 +213,14 @@ case class AtomicReplaceTableAsSelectExec(
  * Rows in the output data set are appended.
  */
 case class AppendDataExec(
-    session: SparkSession,
     table: SupportsWrite,
-    relation: DataSourceV2Relation,
     writeOptions: CaseInsensitiveStringMap,
-    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+    query: SparkPlan,
+    afterWrite: () => Unit = () => ()) extends V2TableWriteExec with BatchWriteHelper {
 
   override protected def run(): Seq[InternalRow] = {
     val writtenRows = writeWithV2(newWriteBuilder().buildForBatch())
-    session.sharedState.cacheManager.recacheByPlan(session, relation)
```

Comment:
I thought about this when I reviewed the previous PR but there are already other occurrences of passing it such as `DropTableExec` and `RefreshTableExec`. Maybe it's better to file a separate JIRA and handle them all.

Should we maybe just pass `session.sharedState.cacheManager`? I see analogical catalog manager is being passed around. I think we can do it for `DropTableExec` and `RefreshTableExec` too.

Reply:
cc @rdblue and @cloud-fan

Reply:
Yeah I'm fine by doing these together in another JIRA. One issue is that the recache logic occurs in various Spark plans right now, and it would be nice to put it in a central place (perhaps via another util trait?) so the code is more manageable.

Reply:
I like the callbacks to do this so that we don't need to pass the spark session, catalog manager, or logical plans into physical plans.

Reply:
What about fixing `DropTableExec` and `RefreshTableExec` together since we're here with a separate JIRA? There look two other DSv2 execution plans left that pass Spark session.

Doing a refactoring across the codebase could be also done separately I believe.

One thing though the callback  always does the same thing but it's unlikely for `afterWrite` to do something else at least so far. I felt it's a bit too much to make it generalised. But probably it's just okay.

Reply:
Yea let's create a new JIRA and fix `DropTableExec` and `RefreshTableExec` together.

Reply:
+1 for a separate JIRA for all callbacks.